core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/parse.kt

```diff
@@ -408,18 +542,44 @@ internal fun <T> DataColumn<String?>.parse(parser: StringParser<T>, options: Par
     return DataColumn.createValueColumn(name(), parsedValues, parser.type.withNullability(hasNulls)) as DataColumn<T?>
 }
 
-internal fun <T> DataFrame<T>.parseImpl(options: ParserOptions?, columns: ColumnsSelector<T, Any?>) =
-    convert(columns).to {
-        when {
-            it.isFrameColumn() -> it.cast<AnyFrame?>().parse(options)
-
-            it.isColumnGroup() ->
-                it.parse(options) { all() }
-                    .asColumnGroup(it.name())
-                    .asDataColumn()
-
-            it.typeClass == String::class -> it.cast<String?>().tryParse(options)
+internal fun <T> DataFrame<T>.parseImpl(options: ParserOptions?, columns: ColumnsSelector<T, Any?>): DataFrame<T> =
+    runBlocking { parseParallel(options, columns) }
+
+private suspend fun <T> DataFrame<T>.parseParallel(
+    options: ParserOptions?,
+    columns: ColumnsSelector<T, Any?>,
+): DataFrame<T> =
+    coroutineScope {
+        val convertedCols = getColumnsWithPaths(columns).map { col ->
+            async {
+                when {
+                    // when a frame column is requested to be parsed,
+                    // parse each value/frame column at any depth inside each DataFrame in the frame column
+                    col.isFrameColumn() ->
+                        col.values.map {
+                            async {
```

Comment:
do we really need a 2-level coroutine here? How many coroutines will be created during parsing large file, saying 1000 000 rows on 100 columns?

Reply:
Probably we need to set up correct dispatcher or give this ability to the user 
https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-dispatcher/

In the case, if we run with the Default Dispatcher, it could consume too many resources on the machine.

Reply:
>do we really need a 2-level coroutine here? How many coroutines will be created during parsing large file, saying 1000 000 rows on 100 columns?

Coroutines are cheap :) DataFrames in the cells of frame columns are independent of each other, as are columns in a DF. It only makes sense to split them off in different coroutine branches.

>Probably we need to set up correct dispatcher or give this ability to the user

I agree, but I'm not sure how to do this neatly from the API. The way to do it correctly, would be to make `parse()` and all its overloads suspend functions. That way the user can decide which scope to run it on and with which dispatcher. The problem is that the DF API is not built around coroutines, nor should users be forced to put every call in a suspending context, so this would require all overloads to be written twice, both with and without `suspend`... @koperagen any ideas?